pkg/splunk/client/azureblobclient.go

```diff
@@ -0,0 +1,238 @@
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// blank assignment to verify that AzureBlobClient implements BlobClient
+var _ RemoteDataClient = &AzureBlobClient{}
+
+// DownloadOptions ...Options for downloading the apps
+type DownloadOptions struct {
+	MaxChunkSize int // chunk size in gb
+	Timeout      int //timeout in seconds
+}
+
+// Config contains configuration for httpclient
+type Config struct {
+	MaxRetries int         // max number of retries for the rest calls
+	HTTPClient http.Client // the http client to use for calling the rest end points
+}
+
+// ListBlobsOptions contains options for listing the apps
+type ListBlobsOptions struct {
+	Prefix      string // prefix of the files to list, useful to filter apps based on the path they belong to
+	MaxEnteries int    // total number of files to return in one call
+}
+
+// ListBlobsOutput contains XML output from the blob rest call
+type ListBlobsOutput struct {
+	BlobsFlatXML string
+}
+
+// SplunkAzureBlobClient is an interface to AzureBlob  client
+type SplunkAzureBlobClient interface {
+	ListApps(ctx context.Context, bucketName string, listAppsOpts map[string]string) ([]byte, error)
+	DownloadApp(ctx context.Context, bucketName string, remoteFileName string, localFileName string, downloadOpts map[string]string) error
+}
+
+// SplunkAzureBlobClientTemp is temp struct to let the build successfull and
+// have a place holder for azureblobclient
+type SplunkAzureBlobClientTemp struct {
+	client http.Client
+}
+
+// ListApps to be revisited
+func (client *SplunkAzureBlobClientTemp) ListApps(ctx context.Context, bucketName string, listAppsOpts map[string]string) ([]byte, error) {
+	return nil, nil
+}
+
+// DownloadApp to be revsited
+func (client *SplunkAzureBlobClientTemp) DownloadApp(ctx context.Context, bucketName string, remoteFileName string, localFileName string, downloadOpts map[string]string) error {
+	return nil
+}
+
+// AzureBlobClient is a client to implement Azure Blob specific APIs
+type AzureBlobClient struct {
+	BucketName         string
+	StorageAccountName string
+	SecretAccessKey    string
+	Prefix             string
+	StartAfter         string
+	Endpoint           string
+	Client             SplunkAzureBlobClient
+}
+
+// NewAzureBlobClient returns an AzureBlob client
+func NewAzureBlobClient(ctx context.Context, bucketName string, storageAccountName string, secretAccessKey string, prefix string, startAfter string, region string, endpoint string, fn GetInitFunc) (RemoteDataClient, error) {
```

Comment:
instead of passing so many arguments should we pass a structure , like `BucketCredentials` or `BucketDetails`

Reply:
Like the idea. Let me revisit this in one of the upcoming sprints if that's ok with you.